src/compiler/parser.ts

```diff
@@ -1324,10 +1325,7 @@ function setExternalModuleIndicator(sourceFile: SourceFile) {
     sourceFile.externalModuleIndicator = isFileProbablyExternalModule(sourceFile);
 }
 
-export function createSourceFile(fileName: string, sourceText: string, languageVersionOrOptions: ScriptTarget | CreateSourceFileOptions, setParentNodes?: boolean, scriptKind?: ScriptKind): SourceFile;
-/** @internal */
-export function createSourceFile(fileName: string, sourceText: string, languageVersionOrOptions: ScriptTarget | CreateSourceFileOptions, setParentNodes?: boolean, scriptKind?: ScriptKind, skipNonSemanticJSDoc?: boolean): SourceFile;
-export function createSourceFile(fileName: string, sourceText: string, languageVersionOrOptions: ScriptTarget | CreateSourceFileOptions, setParentNodes = false, scriptKind?: ScriptKind, skipNonSemanticJSDoc?: boolean): SourceFile {
+export function createSourceFile(fileName: string, sourceText: string, languageVersionOrOptions: ScriptTarget | CreateSourceFileOptions, setParentNodes = false, scriptKind?: ScriptKind, jsDocParsingMode?: JSDocParsingMode): SourceFile {

```

Comment:
Since we're changing it anyway, should we just fold `jsDocParsingMode` into `CreateSourceFileOptions`?

Reply:
My original PR worked that way, yeah, I just found it a little irritating to have to `{ ...options, jsDocParsingMode }` all over the place in order to set it. Let me recheck how bad that is.

Reply:
The difference is that `languageVersionOrOptions` is basically always set by the caller, whereas `setParentNodes` is always provided by the host more or less like `jsDocParsingMode`. But I also don't like how many functions have like 50 parameters 😄

Reply:
Hm, this gets very hairy. Providing an object instead of a version causes behavior changes, so I can't override it for everything. And if I don't override it and instead attempt to figure out the call chain up to who passes in the object or version in the first place, I'm forced to change the behavior and/or plumb it in an awkward, more-error-prone way...

Reply:
> Providing an object instead of a version causes behavior changes

Whu - it does? AFAIK `{ languageVersion: ts.ScriptTarget.Latest }` and `ts.ScriptTarget.Latest` as parameters should behave identically. What changes?

Reply:
`setExternalModuleIndicator` and `impliedNodeFormat` differ, as of the PR where you introduced that stuff. It's possible that it doesn't matter but I think there's a reason we left the old signature behind, and it's still used heavily in services.

Reply:
I think I'm just totally wrong, apologies!!

https://github.com/microsoft/TypeScript/commit/12f59036b948188126e117ca7510703452462f4d shows that this would theoretically work, and I can use the current mechanism to set it as a default instead and not plumb it all the way out.